cmd/snap-seccomp-blacklist/snap-seccomp-blacklist.c

```diff
@@ -0,0 +1,215 @@
+        }
+    }
+
+    /* When the rule set doesn't match one of the architectures above then the
+     * resulting action should be a "allow" rather than "kill". We don't add
+     * any of the 32bit architectures since there is no need for any extra
+     * filtering there. */
+    sc_err = seccomp_attr_set(ctx, SCMP_FLTATR_ACT_BADARCH, SCMP_ACT_ALLOW);
+    if (sc_err < 0) {
+        showerr("cannot set action for unknown architectures");
+        return sc_err;
+    }
+
+    /* Resolve the name of "ioctl" on this architecture. */
+    int sys_ioctl_nr;
+    sys_ioctl_nr = seccomp_syscall_resolve_name("ioctl");
+    if (sys_ioctl_nr == __NR_SCMP_ERROR) {
+        showerr("cannot resolve ioctl system call number");
+        return -ESRCH;
+    }
+
+    /* All of the rules must be added for the native architecture (using native
+     * system call numbers). When the final program is generated the set of
+     * architectures added earlier will be used to determine the correct system
+     * call number for each architecture.
+     *
+     * In other words, As such arguments to scmp_rule_add() must always use
+     * native system call numbers. Translation for the correct architecture
+     * will be performed internally. This is not documented in libseccomp. But
+     * correct operation was confirmed using the pseudo-code program and the
+     * bpf_dbg tool from the kernel tools/bpf directory.
+     *
+     * NOTE: I'm not using scmp_rule_add_exact as that was not doing anything
+     * at all (presumably due to having all the architectures defined). */
+
+    const struct scmp_arg_cmp no_tty_inject = {
+        /* When faced with ioctl(fd, request); refuse to proceed when
+         * request&0xffffffff == TIOCSTI. This specific way to encode the
+         * filter has the following important properties:
+         *
+         * 1) it blocks ioctl(fd, TIOCSTI, ptr).
+         * 1) it also blocks ioctl(fd, (1UL<<32) | TIOCSTI, ptr).
+         * 2) it doesn't block ioctl(fd, (1UL<<32) | (request not equal to TIOCSTI), ptr);
+         *
+         * We learned that existing programs make legitimate requests with all
+         * bits set in the more significant 32bit word of the 64 bit double
+         * word. While this kernel behavior remains suspect and presumably
+         * undesired it is unlikely to change for backwards compatibility
+         * reasons. As such we cannot block all requests with high-bits set. */
+        .arg = 1,
+        .op = SCMP_CMP_MASKED_EQ,
+        .datum_a = 0xffffffffUL,
+        .datum_b = TIOCSTI,
+    };
+    sc_err = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), sys_ioctl_nr, 1, no_tty_inject);
+
+    if (sc_err < 0) {
+        showerr("cannot add rule preventing the use high bits in ioctl");
+        return sc_err;
+    }
+    return 0;
+}
+
+typedef struct arch_set {
+    const char *name;
+    const uint32_t *arch_tags;
+    size_t num_arch_tags;
+} arch_set;
+
+int main(int argc, char **argv) {
```

Comment:
Out of curiosity, couldn't this have been written in Go?

Reply:
Only in the go that imports a preamble of C headers. I strongly desire to stop using C for this code but using it as a super thin layer on top of C is not a solution IMO.